lib/hlclock.ex

```diff
@@ -1,50 +1,104 @@
 defmodule HLClock do
   @moduledoc """
-  Documentation for HLClock.
+
+  Hybrid Logical Clock
+
+  Provides globally-unique, monotonic timestamps. Timestamps are bounded by the
+  clock synchronization constraint, max_drift.
+
+  Implementation assumes that timestamps are (at a minimum) regularly sent; a
+  clock at rest will eventually be unable to generate timestamps due to assumed
+  bounds on the logical clock relative to physical time.
+
+  In order to account for physical time drift within the system, timestamps
+  should regularly be exchanged between nodes. Generate a timestamp at one node
+  via HLClock.send_timestamp/1; at the other node, call HLClock.recv_timestamp/2
+  with the received timestamp from the first node.
+
+  Inspired by https://www.cse.buffalo.edu/tech-reports/2014-04.pdf
   """
 
+  alias __MODULE__, as: Clock
   alias HLClock.Timestamp
 
-  def new(pt \\ default_clock()) do
-    Timestamp.new(pt, 0)
+  defstruct [:clock_fn, :curr_ts]
+
+  @doc """
+  clock constructor requires the node_id, a millisecond clock fn and a
+  maximum drift parameter in milliseconds
+  """
+  def new(node_id \\ 0,
+    clock_fn \\ &Clock.default_time/0) do
+    with {:ok, t0} <- Timestamp.new(clock_fn.(), 0, node_id) do
+      %Clock{clock_fn: clock_fn,
+             curr_ts: t0}
+    end
   end
 
-  def now(clock, pt \\ default_clock()) do
-    if Timestamp.ahead?(clock, pt) do
-      Timestamp.increment(clock)
-    else
-      Timestamp.new(pt, 0)
+  @doc "generate a new timestamp"
+  def send_timestamp(%Clock{curr_ts: old, clock_fn: cfn} = clock) do
+    pt = cfn.()
+    nt = max(old.time, pt)
+    nc = advance(old, nt)
+
+    cond do
+      drift?(clock, nt, pt) ->
+        {:error, :clock_drift_violation}
+      true ->
+        # counter overflow is handled by Timestamp.new
+        with {:ok, new_ts} <- Timestamp.new(nt, nc, old.node_id) do
+          {:ok, %Clock{clock | curr_ts: new_ts}}
+        end
     end
   end
 
-  def less?(clock1, clock2) do
-    Timestamp.compare(clock1, clock2) == :lt
+  defp advance(old, new_time) do
+    cond do
+      old.time == new_time ->
+        old.counter + 1
+      true ->
+        0
+    end
   end
 
-  def update(local, foreign, pt \\ default_clock()) do
-    max_pt = max(pt, max(local.time, foreign.time))
+  @doc "receive a remote timestamp and merge with local"
+  def recv_timestamp(%Clock{clock_fn: cfn, curr_ts: local} = clock, msg) do
```

Comment:
I think I'm going to drop the struct. `msg` is a timestamp, but I'll go with something like remote vs local.